### PR TITLE
[REG-1913] Open vocab text query object detection bot segment criteria. 

### PR DESCRIPTION
**Pull request by batu**
[Loom video](https://www.loom.com/share/7d55604f8e2c445daabbf32dab9569c5?sid=613050f0-da98-48c3-9078-fcb358121575).

Implemented the open vocab text query bot segment criteria. 

[Example segments on Boss Room.](https://github.com/Regression-Games/RGBossRoom/pull/67)

There are some follow up tasks that are related to this:
- [Refactor this and CV Image match evaluators.](https://linear.app/regression-games/issue/REG-1928/refactor-cv-object-detection-evaluator)
- [Update Python endpoints to match.](https://linear.app/regression-games/issue/REG-1920/update-python-object-detection-endpoints-to-match-sdk)

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+        public void WriteToStringBuilder(StringBuilder stringBuilder)
+        {
+
+            bool textQuery = queryText != null;
+            bool imageQuery = queryImage != null;
+
+            if (textQuery && imageQuery)
+            {
+                RGDebug.LogError("You need to use either textQuery or imageQuery. You can not have both.");
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs:39]: Does our WriteToStringBuilder approach throw errors in other instances? Trying to determine if this should be caught here or elsewhere.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+using RegressionGames.StateRecorder.BotSegments.JsonConverters;
+using RegressionGames.StateRecorder.JsonConverters;
+
+namespace RegressionGames.StateRecorder.BotSegments.Models
+{
+    [Serializable]
+    public class CVObjectDetectionKeyFrameCriteriaData : IKeyFrameCriteriaData
+    {
+        // update this if this schema changes
+        public int apiVersion = SdkApiVersion.VERSION_9;
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs:14]: Is this the latest version? There may be a newer one, not sure.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        public string text;
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs:16]: Let's add a comment on what this is used for. Same for other class variables.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        /**
+       * base64 encoded byte[] of jpg image data , NOT the raw pixel data, the full jpg file bytes
+       */
+        // public string imageData; This will be added in the next PR.
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs:20]: If there is a ticket for this, we usually do the following so we know:

```suggestion
        // TODO(REG-???): Supporting image queries will be added soon
        // public string imageData;
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs [hunk 5]
```diff
@@ -193,8 +197,44 @@ internal bool MatchedHelper(bool firstSegment, int segmentNumber, bool botAction
...
+                    // cvText results will be null until we get the async response back
+                    if (andOr == BooleanCriteria.And)
+                    {
+                        _newUnmatchedCriteria.Add("Waiting for CV Object Detection evaluation results ...");
+                        return false;
+                    }
+                }
+                else
+                {
+                    var cvObjectDetectionResultsCoount = cvObjectDetectionResults.Count;
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs:215]:
```suggestion
                    var cvObjectDetectionResultsCount = cvObjectDetectionResults.Count;
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs [hunk 6]
```diff
@@ -193,8 +197,44 @@ internal bool MatchedHelper(bool firstSegment, int segmentNumber, bool botAction
...
+                        {
+                            if (andOr == BooleanCriteria.And)
+                            {
+                                _newUnmatchedCriteria.Add(resultEntry);
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs:236]: Assuming this is basically a copy-paste of the logic for the other CV stuff? If so, 👍

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+using RegressionGames.StateRecorder.BotSegments.Models;
+using RegressionGames.StateRecorder.BotSegments.Models.CVService;
+using RegressionGames.StateRecorder.JsonConverters;
+using UnityEngine;
+
+namespace RegressionGames.StateRecorder.BotSegments
+{
+
+    /**
+     * <summary>Evaluates CV Image criteria using CVServiceManager to send/receive HTTP requests to a python server for doing the actual CV evaluations.
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs:14]:
```suggestion
     * <summary>Evaluates CV Object criteria using CVServiceManager to send/receive HTTP requests to a python server for doing the actual CV evaluations.
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+     * Python 'detects' a source image template in a provided screenshot and then this class evaluates those results against the specified bot segment criteria.</summary>
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs:15]:
```suggestion
     * Python 'detects' classes of objects in a provided screenshot based on image and text queries, and then this class evaluates those results against the specified bot segment criteria.</summary>
```

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+            var resultList = new List<string>();
+
+            // If no results and no request, we'll initiate a new request.
+            if (!requestInProgress && objectDetectionResults == null)
+            {
+                resultList = TryPostRequest(segmentNumber, criteriaList, resultList);
+            }
+            // If we have results and no request in progress, we'll evaluate them.
+            // We check priorResults null here so that we only evaluate a new result 1 time.
+            else if (!requestInProgress && objectDetectionResults != null && priorResults == null)
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs:120]: I noticed that this check was slightly different from the check in the other file. Just curious to learn more about that!

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+        /// <param name="objectDetectionResults">The list of object detection results for the segment.</param>
+        /// <param name="resultList">The list of result strings to be updated.</param>
+        /// <returns>A list of non-matched entries as strings.</returns>
+        /// <remarks>
+        /// This method performs the following steps:
+        /// 1. Attempts to retrieve existing results for the given segment.
+        /// 2. If no request is in progress and no results exist, it initiates a new request.
+        /// 3. If results are available and no request is in progress, it evaluates the results against the criteria.
+        /// 4. Updates the prior results tracker with the current result list.
+        /// </remarks>
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs:154]: Really really really appreciate all of the comments in this file!

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs (hunk 9 above)
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/CVObjectDetectionEvaluator.cs:120]: I see, this check is in the other method now. Thanks!

**Review by vontell — CHANGES_REQUESTED** [review]
Honestly, amazing amazing work. I know that code could be difficult to sift through, this PR got in much faster than I thought it would. Really nice work! Just a few small comments really.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs (hunk 2 above)
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs:14]: That definitely was not the newest one. Updated to the newest of 15.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs (hunk 1 above)
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs:39]: No, you are definitely right. This is the wrong place to do it. I think we should do it in the constructor

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs (hunk 6 above)
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs:236]: This is a copy paste. I don't fully understand this yet, but I am getting there.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs (hunk 1 above)
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs:39]: I thought about this a little more. This is better but not good enough. Because it is likely the constructor will never actually receive both of these values, Especially if they are set as they are now, like so:

```
                    _ = CVServiceManager.GetInstance().PostCriteriaObjectTextQuery(
                        new CVObjectDetectionRequest(
                            screenshot: new CVImageBinaryData()
                            {
                                width = width,
                                height = height,
                                data = imageData
                            },
                            queryText: queryText,
                            queryImage: null,
                            withinRect: null,
                            index: 1
                        ),
```

Where this check needs to live is in parsing from the json. If in the criteria match json they give both text AND image they should be notified then and there. I added a TODO and the corresponding task.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs (hunk 4 above)
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVObjectDetectionKeyFrameCriteriaData.cs:20]: Done

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionResult.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
+﻿using System.Text;
+using RegressionGames.StateRecorder.JsonConverters;
+using UnityEngine;
+
+namespace RegressionGames.StateRecorder.BotSegments.Models.CVService
+{
+    public class CVObjectDetectionResult
+    {
+        public string text;
```
[the remark this answers is not in the document]
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionResult.cs:9]: This is very wrong. Fixing now, however the fact I missed it is worrying and potentially calls for a hardening.

Basically this is what happened:
1) I assumed compiler would complain about a mismatch between result json and string, when it didn't I didn't double check. This is not the case, as we use 'TryGetValue' to fill in the results, which does its best to load results but doesn't enforce a match.
```c#
public bool TryGetValue(TKey key, out TValue value)
    {
      int entry = this.FindEntry(key);
      if (entry >= 0)
      {
        value = this._entries[entry].value;
        return true;
      }
      value = default (TValue);
      return false;
    }
```
2) The criteria only cares about existence of a match, not whether the data returned is correct (well because we don't have ground truth). So even if the type was very wrong, the behavior worked as expected. This would have fallen over very hard at the action phase.

It is likely there is a reason why this is the way it is, so I am adding a link to the future [REG-1928](https://linear.app/regression-games/issue/REG-1928/refactor-cv-object-detection-evaluator) to discuss hardening

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs (hunk 1 above)
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs]: Ended up fixing it in this PR. However, I have a question about logging errors:
image

With both 
```
RGDebug.LogError("Both textQuery and imageQuery cannot be provided simultaneously. Use only one.");
throw new ArgumentException("Both textQuery and imageQuery cannot be provided simultaneously. Use only one.");
```
I feel like we are duplicating information? Should RGDebug.Log actually raise an error?

Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs]: yeah I'm looking through our code and I am confused about how exactly the pattern would work, I can't find any spot where we throw an error... it seems like we always just terminate, stop loops, etc... 

I suppose putting it into that CVObjectDetectionKeyFrameCriteriaData parsing file is fine... there is an argument to be made that this is not an error for the JSON parsing and is more of an error for the business logic later on... but it gets weird since we are defining this data format and parsing the JSON in the same class, so the class could impose the restrictions I suppose... this seems fine to me!

TLDR; yeah I think this is all good for now, but let's make a note for @RG-nAmKcAz to check this out when he gets back in case he has a pattern in mind that we were following.

Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionRequest.cs]: (mark this as resolved when you read it)

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionResult.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/CVService/CVObjectDetectionResult.cs]: Good catch!

**Review by vontell — COMMENTED** [review]
Just one more small request related to API version - otherwise, looking good!

**Review by vontell — APPROVED** [review]

**Review by addisonbgross — APPROVED** [review]
Pending Aaron's comments, this looks reasonable to me!
